corehq/apps/analytics/ab_tests.py

```diff
@@ -1,53 +1,128 @@
+        :param slug: (string) a unique slug, preferably with a date
+            (my_test_dec2018). Used only in generating the cookie and cache IDs.
+        :param options: (tuple) a tuple of option values
+
+        Optional for DEV USE ONLY:
+        :param is_debug: (boolean) whether debug (logging) mode is active
+        :param force_refresh: (boolean) whether to force a refresh of the values
+        """
+        self.name = name
+        self.slug = slug
+        self.options = options
+        self.is_debug = is_debug
+        self.force_refresh = force_refresh
+
+
+class SessionAbTest(object):
     """
-    Initialize A/B test with an ABTestConfig.
+    Initialize a user session A/B test from a SessionABTestConfig.
+    This type of A/B test is useful for anonymous users (pre login),
+    and uses the users's unique session id and cookies to maintain consistency.
     """
 
     def __init__(self, config, request):
-        self.name = config.name
-        self.slug = config.slug
-        self.options = config.options
+        """
+        :param config: an instance of SessionABTestConfig
+        :param request: view request
+        """
+        self.config = config
         self.request = request
 
     @property
     def _cookie_id(self):
-        return "{}_ab".format(self.slug)
+        return "{}_ab".format(self.config.slug)
+
+    @property
+    def _cache_id(self):
+        return "{}_{}".format(self._cookie_id, self.request.session.session_key)
 
     @memoized
     def version(self, assign_if_blank=True):
-        if self._cookie_id in self.request.COOKIES:
-            return self.request.COOKIES[self._cookie_id]
-        if assign_if_blank:
-            return random.choice(self.options)
-        return None
+        value = None
+        if cache.get(self._cache_id):
+            value = cache.get(self._cache_id)
+            self._debug_message("fetched version from cache '{}'".format(value))
+        elif (self._cookie_id in self.request.COOKIES
+              and not self.config.force_refresh):
+            value = self.request.COOKIES[self._cookie_id]
+            self._debug_message("fetched version from cookie '{}'".format(value))
+        elif assign_if_blank:
+            value = random.choice(self.config.options)
+            self._debug_message("fetched new version '{}'".format(value))
+        self.cache_version(value)
+        return value
+
+    def cache_version(self, version):
+        self._debug_message("cache version '{}' under '{}'".format(
+            version, self._cache_id))
+        cache.set(self._cache_id, version)
+        # fyi default cache timeout is 300 sec (5 min)
+
+    def _clear_cache(self):
+        self._debug_message("clearing cache '{}'".format(self._cache_id))
+        cache.delete(self._cache_id)
+
+    def _debug_message(self, message):
+        if self.config.is_debug:
+            print("SESSION AB TEST DEBUG [{}]: {}".format(self._cookie_id, message))
```

Comment:
Where do these messages go?

Reply:
in the console. intended for debugging on a dev environment

Reply:
aha, cool